### PR TITLE
validate admission-control startup param

### LINKED ISSUE #56614: should validate admission plugin names --admission-control
**Is this a BUG REPORT or FEATURE REQUEST?**:

 /kind bug

**What happened**:
`--admission-control` param is not validated in func Validate 

**What you expected to happen**:

validate admission plugins, if not registered, return err to prevent going on until call `AdmissionOptions.ApplyTo`. 

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

validate admission plugins, if not registered, return err to prevent going on until call AdmissionOptions.ApplyTo.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56614

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 1]
```diff
@@ -129,7 +129,22 @@ func (a *AdmissionOptions) ApplyTo(
 }
 
 func (a *AdmissionOptions) Validate() []error {
+	if a == nil {
+		return nil
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Should not return nil, which will raise exception. `len(nil)` is invalid.

Move `errs := []error{}` up and return `errs` instead.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 2]
```diff
@@ -129,7 +129,22 @@ func (a *AdmissionOptions) ApplyTo(
...
+
 	errs := []error{}
+	registeredPluginsMap := make(map[string]bool)
+	registeredPlugins := a.Plugins.Registered()
+	for _, name := range registeredPlugins {
+		registeredPluginsMap[name] = true
+	}
+
+	for _, name := range a.PluginNames {
+		if registeredPluginsMap[name] != true {
```
Comment (dixudx) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Better to use `if !registeredPluginsMap[name]`.

**Review by dixudx — CHANGES_REQUESTED** [review]
others lgtm

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: make sense

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 2 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: ok

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: I just write a demo . Test result show this style is ok.

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Actually should remove `if a== nil`. It won't be nil. Right?

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: return value is not just nil.  It is an empty slice.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]:
```
// Validate checks invalid config combination
func (o *AuditOptions) Validate() []error {
	if o == nil {
		return nil
	}

	allErrors := []error{}
```
This is a safe way.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Prevent if just declare a var without initialize may panic.
```
var opt *AdmissionOptions
opt.Validate()
```

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: I mean you should remove

```
if a== nil {
xxxx
}
```

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: > return value is not just nil. It is an empty slice.

It's a typed nil of slice type ;-)

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: > I mean you should remove

I think it's fine and correct to keep.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 3]
```diff
@@ -129,7 +129,23 @@ func (a *AdmissionOptions) ApplyTo(
...
+
+	registeredPluginsMap := make(map[string]bool)
+	registeredPlugins := a.Plugins.Registered()
+	for _, name := range registeredPlugins {
+		registeredPluginsMap[name] = true
+	}
+
+	for _, name := range a.PluginNames {
+		if !registeredPluginsMap[name] {
+			errs = append(errs, fmt.Errorf("--admission-control plugin %q is invalid", name))
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: this is not necessarily configured through the flag. Better just call it `admission plugin`.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: make sense

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: a.PluginNames here is all through --admission-control flag.  Validate called before ApplyTo.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Professional name typed nil.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Downstream projects might set up the options without flags and run the validation on them, e.g. OpenShift.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: Understand.

**Comment by hzxuzhonghu** [comment]
ping @mikedanese @nikhiljindal, do you have any comments?

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 4]
```diff
@@ -129,7 +129,23 @@ func (a *AdmissionOptions) ApplyTo(
 }
 
 func (a *AdmissionOptions) Validate() []error {
+	if a == nil {
+		return nil
+	}
+
 	errs := []error{}
+
+	registeredPluginsMap := make(map[string]bool)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: use the string set https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/sets/string.go#L30

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: ok

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 5]
```diff
@@ -129,7 +130,22 @@ func (a *AdmissionOptions) ApplyTo(
 }
 
 func (a *AdmissionOptions) Validate() []error {
+	if a == nil {
+		return nil
+	}
+
 	errs := []error{}
+
+	registeredPluginsMap := sets.NewString()
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: `sets.NewString(a.Plugins.Registered()...)`

**Comment by hzxuzhonghu** [comment]
@sttts pr squashed.

**Pull request merged**